mountainGlacier added

SanderKeulers opened the pull request.

SanderKeulers committed:
copied lithosphereExtension

SanderKeulers committed:
Update material_properties.txt

Changing viscosity of air from 1e18 to 1e-5 will likely not be a very nice change

SanderKeulers committed:
Updated some material properties

SanderKeulers committed:
created topography_curve and ice_curve

SanderKeulers committed:
added some comments

csummers25 commented on models/mountainGlacier/material_properties.txt:
I think that there shouldn't be any radiogenic heating (i.e. heat production from radioactive sources) present in either the ice or the bedrock, so the last value in each row should be zero.

csummers25 commented on models/mountainGlacier/setup.py:
```diff
@@ -0,0 +1,293 @@
+            
+            # glacier
+            if markers.y[mm] >= topography_curve(markers.x[mm]) and markers.y[mm] < ice_curve(markers.x[mm]):
+               markers.id[mm]  = 6
```
The ID numbers start counting from 0 and don't include the comments in the material_properties file, so I think this should be 1, and then air = 0, bedrock = 2

csummers25 commented on models/mountainGlacier/setup.py:
```diff
@@ -0,0 +1,293 @@
+
+    B_bottom = np.zeros((xnum+1,4))
+    B_bottom[:,1] = 1
+    B_bottom[:,2] = -params.v_ext/xsize * ysize
```
I think you will need to take out the grid extension from here, I guess just setting it to no-slip since the bottom shouldn't be moving anyway

csummers25 commented on models/mountainGlacier/setup.py:
```diff
@@ -0,0 +1,293 @@
+    B_bottom[:,2] = -params.v_ext/xsize * ysize
+
+    B_left = np.zeros((ynum+1,4))
+    B_left[:,0] = -params.v_ext/2
```
Same for these boundaries

SanderKeulers commented on models/mountainGlacier/setup.py:
```diff
@@ -0,0 +1,293 @@
+    ysize = ysize0
+
+    # set resolution
+    xnum = 1610
```
Yes, I thought to decrease grid size (were not modelling hundreds of km of glacier) and simultaneously increase the grid resolution. The flow of the glacier will in the end be faster than that of plate tectonics but I'll restore it to the original for now

SanderKeulers committed:
tried to set up grid

grid somehow stops at 1/3 and air temp is not getting in correctly

SanderKeulers committed:
grid with sloping glacier

SanderKeulers committed:
minor changes

SanderKeulers commented:
![Screenshot 2025-05-28 095753](https://github.com/user-attachments/assets/fa935857-f926-4d07-abf9-80d7c2552889)
Created a sloping glacier on the grid, which now looks something like this. 
We do need to take a look at the grid size and resolution because 400 km by 300 km is way too much for a realistic mountain glacier. I guess something like 1-2 km over the vertical and perhaps 10-20 km in the x-direction. 
Running this setup, however, gives some errors

csummers25 commented:
So the physical grid size (`xsize`,`ysize`) is not dependent on the resolution (`xnum`,`ynum`), we can have a 10-20km grid with the same number of grid cells as the lithosphere extension model (xnum=161,ynum=61), we just need to make sure we set the max timestep `params.tstp_max` to a sensible value for the scale of our problem, so that the ice is moving less than a grid cell in a timestep. 

The other thing that's missing is that the parameters for the gridSpacings calculations need overwriting for the new grid size/setup.  For now, I think we should just use a uniform grid so you can copy the lines where we set the parameters from the Simple Stoke's example: https://github.com/csummers25/uu-mod-earth-systems/blob/e9bd9eb76b25cb78dca154d5dfced2b2a61e86fa/models/simpleStokesTest/setup.py#L93

SanderKeulers committed:
Update setup.py

changed bedrock and ice curve, changed back to uniform grid like in the Simple Stoke's

csummers25 commented on models/mountainGlacier/setup.py:
```diff
@@ -0,0 +1,308 @@
+    BT_bottom[:,0] = params.T_bot
+
+    # left and right = insulating?
+    # BT_left[:,1] = 1
```
I think this should be uncommented, without it you are fixing the boundaries at 0.  With this included, the boundary temperature is just the same as the temperature in the first grid cell, which should be ok for now.

SanderKeulers committed:
changed Ad to 1e-24

SanderKeulers commented:
So for the value of A I find values in the order of 10^-24 
![image](https://github.com/user-attachments/assets/cfd65968-8398-4f1e-960e-1e6b7a437b48)
![image](https://github.com/user-attachments/assets/4de4dae1-c7ca-4a23-af10-7b7903208b2d)

which we can take as constant for now, but it apparently follows Arrhenius law quite reasonably, so we may implement this at some point 
![image](https://github.com/user-attachments/assets/2edc0e88-1fa9-48e2-b563-31079fef4399)

And in the material properties text file, what is Ea and what value should this be for a glacier?

csummers25 commented:
So in the code we have the viscosity given by $A_d \sigma^n exp\left(-\frac{E_a + V_a P}{RT}\right)$ where $\sigma$ is the stress (which is the same as $\tau$ in the Arrhenius law).  So I think that comparing the two formulae, 
 - $A_d$ is $A_0$
 - $n$ is $n-1$ 
 - $E_a$ is $Q$, and if you don't want temperature dependence then it should be 0.
 - $V_a$ is 0 since there's no pressure dependence in the model.

SanderKeulers committed:
Update material_properties.txt

csummers25 commented:
Ok, it looks like the reason it is stuck in the temperature solver is that the RHS of the temperature eqns have some very large values.  This is coming from some problems with the viscosity model, so you'll find that if you switch off the friction heating term by setting`params.frict_yn = 0` you no longer get stuck in the temperature solver but crash after a few timesteps.

It seems that the power law you're using for the ice is producing a viscosity above the maximum set by `params.max_eta` so it is just being set to that value (`1e25` as a default).  You're also trying to set a realistic viscosity for the air, but you haven't changed the `params.min_eta` from its default of `1e18` so the air is being set to that.  

What you should do here is to try and fix the ice viscosity model so that it gives a correct value (maybe even just set it to a constant that's approximately right for now?), then set the air viscosity to be something lower than it, but not its realistic value since the code cannot handle such a large change in viscosity.  The same for the rock in the other direction, then I would set the `min_eta` and `max_eta` values to the air and rock viscosities respectively.  If you aim for something like the order of magnitude range present in the lithosphere code (about 7 orders of magnitude), but centered on the ice viscosity then it should stay stable, but you might have to narrow the range further.
